ADR: Resource based API

ch1bo opened the pull request:
An approach to compartmentalize the API into smaller parts to make them available for a wider range of use cases.

Inspired by previous experience on using HTTP methods for commands and HTTP upgraded connections for subscriptions.

Also very similar to GraphQL. 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Early draft for a resource based API

ch1bo committed:
Updated early draft with more concrete proposals

pgrange commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+  - Need to poll `GetUTxO` _or_ aggregate confirmed transactions on client side
+    to know the latest UTxO for constructing transactions.
+
+  - Inclusion of the whole UTxO in the head is not always desirable and
```
```suggestion
  - Inclusion of the whole UTxO set in the head is not always desirable and
```

pgrange commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+    very basic `headStatus` from the `Greetings`).
+
+  - Need to poll `GetUTxO` _or_ aggregate confirmed transactions on client side
+    to know the latest UTxO for constructing transactions.
```
```suggestion
    to know the latest UTxO set for constructing transactions.
```

pgrange commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+| Path                        | GET                         | POST                   | PATCH   | DELETE             |
+| :-------------------------- | :-------------------------- | :--------------------- | ------- | :----------------- |
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
+| `/node/head/utxo`           | confirmed utxo              | -                      | -       | -                  |
```
I'm guessing _utxo_ is meant for _utxo set_ here.

If I'm correct, what would mean _confirmed utxo_? Is it for the utxo that are included in a snapshot?

Also, I've noticed we currently have some issues with the clients only having access to the utxo included in a snapshot (see #612 ). I feel like `/node/head/utxo`, if containing the set of utxo included in a snapshot would exhibit the same problem and I would like us to solve it before. In my opinion, the client should, by default, only access the latest view of its node ledger when constructing a transaction. Whether or not these utxo have been included in a snapshot should not be considered at this step.

Last, but not least, we can currently have issues when the utxo set is very big and the _snapshot confirmed_ message is heavy and slows down the node. That led some users to ask for utxo query primitives instead of getting the full utxo set. I'm mentioning this, although I'm not sure we want to address it right now as having a huge utxo set would make it impossible to close the head with the current implementation anyway.

ch1bo replied:
Good points.

We usually mean any number of unspent transaction outputs when we just say `utxo` or `UTxO`. This is also how it is documented right now in the api reference and glossary IIRC.

I commented below that we likely need both, local and confirmed versions of utxo and transactions.

I will add a section on how the querying of such big resources could be improved using content types and query parameters (and maybe even pagination?). But maybe I also don't want to take a decision on that in this ADR (it's rather a consequence) and ultimately we might not be providing for this and suggest indexers instead?

pgrange commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+| :-------------------------- | :-------------------------- | :--------------------- | ------- | :----------------- |
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
+| `/node/head/utxo`           | confirmed utxo              | -                      | -       | -                  |
+| `/node/head/transactions`   | confirmed txs               | `NewTx` + responses    | -       | -                  |
```
Same question: does _confirmed txs_ mean they have been included in a snapshot?

ch1bo replied:
Yes, that is what I meant. I had `Snapshot{confirmed}` here before. Maybe that would be more clear?

pgrange commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
+| `/node/head/utxo`           | confirmed utxo              | -                      | -       | -                  |
+| `/node/head/transactions`   | confirmed txs               | `NewTx` + responses    | -       | -                  |
+| `/node/head/ledger`         | `localUTxO` and `localTxs`  | -                      | -       | -                  |
```
Now I think I understand the following:

* `/node/head/utxo` -> utxo set of the last known validated snapshot
* `/node/head/transactions` -> transactions included in the last known validated snapshot (transitively)
* `/node/head/ledger`  -> utxo set and transactions of the local ledger, irrespective of the snapshots

I'm not sure we want to expose so much details to the customer. This notion of local transactions and local utxo set feels very internal to the Hydra protocol specification/implementation.

I think, as a client of Hydra, that I want to be able to :
1. init/open/close/contest/fanout a head
2. build new L2 transactions
3. know about the finality of a transactions on L2

For 1. I need the most recent view of the ledger from my node. Otherwise, I could build locally invalid transactions (see #612 ) so `/node/head/ledger` would be enough, although I would name it differently.

For 2. I need the information that a given transaction has been included in a signed snapshot and, for that, I guess I would want to have an entry point to get the status of the transaction itself and it would inform me about the fact that it's _pending_ and _snapshotted_.

That would lead to the following entry points:

* `/node/head/utxo` -> utxo set known by my local node
* `/node/head/transactions` -> transactions known by my local node
* * `/node/head/transactions/<id>` -> informations about transaction of id _id_, can be used to check for L2 finality of a transaction
* `/node/head/ledger`  -> removed

ch1bo replied:
I would say exactly the other way around?

For 1) you would care about which UTxO are available on the L2 after opening and which would be available if we close (any non confirmed utxo would not be included!)

For 2) you would need the most recent local view of the UTxO set, such that your node accepts transactions moving this forward.

For 3) we can do this individual query as you propose, but I would advise against it as it does not allow us to setup a subscription for all confirmed L2 transactions, which is a use case we know exists.As you might 

So we are in the situation where likely need both, the local view and the confirmed view. I will try to make that separation more clear in the example resources and also should make it clear that the ADR is not primarily about specifying the full API, but more to exercise the concept of mapping endpoints to existing semantics and see whether we could address the requirements from the context.

pgrange reviewed (approved):
All in all very clear proposal. I approve in draft status.

Please take a look at the questions in my comments.

v0d1ch commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+
+## Decision
+
+- Do not add a general purpose querying interface (e.g. graphql) to the
```
I assume you don't want to add any persistence a part from the one we have right now (dump to file) so how are these query models suppose to work? Would you add different file for different models? If not how would you get the latest known state for each model? Seems like you would need to search the file to get the output you are looking for or perhaps I am missing something?

v0d1ch replied:
Ok, I see. You want to use projections to do this.

ch1bo replied:
I just realize.. what this ADR evolved into, is exactly what graphql could provide. The [graphql subscriptions](https://www.apollographql.com/docs/react/data/subscriptions/) are non-standard though, but work basically the same as described in this ADR

ch1bo replied:
I removed this statement.

v0d1ch commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+    binary (`cbor`) output of `transaction` fields in several of these using a
+    query parameter.
+
+- Many of these features have been added in a "quick and dirty" way, by monkey
```
Does this adr assume also changing the "quick and dirty" way of doing things?

ch1bo replied:
Yes. I added some notes on how content types and query parameters can lead to a fine grained control and more standard way of configuring responses.

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+
+| Path                        | GET                         | POST                   | PATCH   | DELETE             |
+| :-------------------------- | :-------------------------- | :--------------------- | ------- | :----------------- |
+| `/node/head/status`         | `HeadStatus(..)`            | -                      | -       | -                  |
```
Good point. The idea here really was to make sub-paths always included in the parent. `/node` would contain all the remaining data available at `/node/*` endpoints. There is also the intention to keep `/` with the same semantics as before.

Shall we
1. Make it an exception to the rule and say `/` will contain not everything in `/*`, but the original behavior of accepting a websocket upgrade with the same messages as before? or
2. Rename `/node` to `/v1` and make `/v1/*` include all the data from sub-paths to make it more consistent?

ch1bo replied:
I opted for doing 2) and would love your feedback again.

ch1bo committed:
Small rewording and use StateChanged to define resources/ServerOutput

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+  - Each resource is only available as `JSON` encoding (for now).
+
+    - (TBD): already specify that `Accept` headers should be used to negotiate a
+      `content-type`?
```
I reworded this in a way that JSON should be right away supported by all resources. Others are optional.

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+
+### Resources
+
+Resource paths + HTTP verbs mapped to existing things:
```
I added a `/v1` instead of the `/node`

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+  into topics on the API layer.
+
+  - Each resource has a _latest_ state, which can be queried or clients may
+    subscribe to any model changes.
```
Are subscriptions in scope of standard graphql? Couldn't see it in most implementations and only apollo supports that.

It's quite inspired by graphql and maybe just doing that is simpler? Happy to discuss!

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,146 @@
+    subscribe to any model changes.
+
+  - A resource's `model` type needs to be a result of a pure `projection` from
+    server output events, i.e. `project :: model -> ServerOutput -> model`.
```
Yes. Updated the wording on this one. Hope it's more clear. Not sure though if projection is the right word here. I also like "reduce" as it may be familiar with most people from the frontend world of redux state management systems.

ch1bo committed:
Rewrite parts of decision section for better clarity

This should make it more clear that content-types and query parameters
should be used to negotiate response structure on a per-resource basis

ch1bo committed:
Version the proposed api and update examples

The included example resources should only be seen as that. Not all
details about separating clearly between local and confirmed utxo/txs
are resolved in this ADR.

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,145 @@
+    filtering by address would be beneficial. (not addressed in this ADR though)
+
+  - As [ADR-15](/adr/15) also proposes, some clients may not need (or should
+    not have) access to administrative information.
```
Indirectly. There is a consequence (number 4) which mentions this again.

ch1bo commented on docs/adr/2023-08-18_025-resource-based-api.md:
```diff
@@ -0,0 +1,145 @@
+| `/v1/`                           | all `/v1/*` data               | -                      | -       | -                  |
+
+Multiple heads are out of scope now and hence paths are not including a
+`<headId>` variable section.
```
I'd like to keep it out for now.

ch1bo committed:
Added more links
